server/src/main/java/org/opensearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -238,10 +239,20 @@ class AsyncAction {
                 assert request.concreteNodes() != null;
             }
             this.responses = new AtomicReferenceArray<>(request.concreteNodes().length);
+
+            if (request.retainDiscoveryNodes() == false) {
+                // We transfer the ownership of discovery nodes to route the request to into the AsyncAction class.
+                // This reduces the payload of the request and improves the number of concrete nodes in the memory
+                this.concreteNodes = request.concreteNodes();
+                request.setConcreteNodes(null);
+            } else {
+                // initializing it separately as we keep the `concreteNodes` as final since we want it to be immutable.
+                this.concreteNodes = null;
```

Comment:
why can't we initialize this.concreteNodes = request.concreteNodes(); irrespective of whether retainDiscoveryNodes() is true or false. It will be release as soon as the request has been sent to all nodes?

Reply:
if check will be for
```
this.concreteNodes = request.concreteNodes();
if (request.retainDiscoveryNodes() == false) {
                request.setConcreteNodes(null);
    }
```
and no need for else

Reply:
Got it. Just wanted to reduce the number of list of discovery nodes in the heap at any point of time. If we set the concrete nodes but not use it, might as well not use. 

But I think this should not cause any regression as such so will change as per suggestion. Thanks!